File: backend/app/schemas/presentation.py
from __future__ import annotations
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class PresentationCreate(BaseModel):
    title: str = Field(default="Untitled Presentation", max_length=500)
    folder_id: uuid.UUID | None = None
    aspect_ratio: str = Field(default="16:9", pattern="^(16:9|4:3)$")
    template_id: uuid.UUID | None = None


class PresentationUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    folder_id: uuid.UUID | None = None
    slides_meta_json: dict | None = None
    theme_json: dict | None = None
    aspect_ratio: str | None = Field(None, pattern="^(16:9|4:3)$")


class PresentationResponse(BaseModel):
    id: uuid.UUID
    title: str
    owner_id: uuid.UUID
    folder_id: uuid.UUID | None = None
    slides_meta_json: dict | None = None
    slide_count: int
    theme_json: dict | None = None
    aspect_ratio: str
    is_template: bool
    template_category: str | None = None
    thumbnail_key: str | None = None
    is_trashed: bool
    last_edited_by: uuid.UUID | None = None
    last_edited_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend-app/src/types/presentation.ts
export interface Presentation {
  id: string;
  title: string;
  owner_id: string;
  folder_id: string | null;
  slides_meta_json: SlidesMeta | null;
  slide_count: number;
  theme_json: PresentationTheme | null;
  aspect_ratio: "16:9" | "4:3";
  is_template: boolean;
  template_category: string | null;
  thumbnail_key: string | null;
  is_trashed: boolean;
  last_edited_by: string | null;
  last_edited_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface SlidesMeta {
  slides: SlideInfo[];
  elements: Record<string, SlideElement[]>;
}

export interface SlideInfo {
  id: string;
  layout: string;
  transition: string;
  transitionDuration: number;
  speakerNotes: string;
  hidden: boolean;
}

export interface SlideElement {
  type: "textbox" | "shape" | "image" | "table" | "chart";
  x: number;
  y: number;
  width: number;
  height: number;
  rotation?: number;
  text?: string;
  fontSize?: number;
  bold?: boolean;
  italic?: boolean;
  color?: string;
  textAlign?: string;
  shape?: string;
  fill?: string;
  stroke?: string;
  strokeWidth?: number;
  src?: string;
  opacity?: number;
}

export interface PresentationTheme {
  primary: string;
  secondary: string;
  accent: string;
  background: string;
  text: string;
  headingFont: string;
  bodyFont: string;
}
